### PR TITLE
Clean expired quotes - Fix out of memory on huge quotes list

### LINKED ISSUE #28342: [Issue] Clean expired quotes - Fix out of memory on huge quotes list
This issue is automatically created based on existing pull request: magento/magento2#27260: Clean expired quotes - Fix out of memory on huge quotes list

---------
### Description (*)
On production we had a lot of expired quotes, and during cron run we were getting fatal error:
```
PHP Fatal error:  Allowed memory size of 2147483648 bytes exhausted (tried to allocate 20480 bytes) in /path/to/magento2/vendor/magento/framework/Model/AbstractModel.php on line 359

Check https://getcomposer.org/doc/articles/troubleshooting.md#memory-limit-errors for more info on how to handle out of memory errors.
```

Investigation found that cron job `sales_clean_quotes` fails with out of memory because it tries to load all quote items at once

### Related Pull Requests
N/A

### Fixed Issues (if relevant)
N/A

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by ihor-sviziev**
### Description (*)
On production we had a lot of expired quotes, and during cron run we were getting fatal error:
```
PHP Fatal error:  Allowed memory size of 2147483648 bytes exhausted (tried to allocate 20480 bytes) in /path/to/magento2/vendor/magento/framework/Model/AbstractModel.php on line 359

Check https://getcomposer.org/doc/articles/troubleshooting.md#memory-limit-errors for more info on how to handle out of memory errors.
```

Investigation found that cron job `sales_clean_quotes` fails with out of memory because it tries to load all quote items at once

### Related Pull Requests
N/A

### Fixed Issues (if relevant)
N/A

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#28342: Clean expired quotes - Fix out of memory on huge quotes list

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
I think it's a good idea to cover these changes with some test, but have no ideas which test will be ok there

**Review by lbajsarowicz — COMMENTED** [review]
If you wait for my pagination fix, your "for" loop with comment would no longer be needed and could be replaced with while.

**Comment by ihor-sviziev** [comment]
Yeah, I know, but it works anyway with new and old version, so I decided to
keep it like that.

**Comment by lbajsarowicz** [comment]
The difference is that with while you continuously get the real amount of entries. With for - you get number of pages once and then iterate

**Comment by ihor-sviziev** [comment]
Hi @lbajsarowicz,
Could you review my changes again? I covered them with integration test

File: dev/tests/integration/testsuite/Magento/Sales/Cron/CleanExpiredQuotesTest.php [hunk 1]
```diff
@@ -66,4 +73,24 @@ public function testExecute()
...
+        //Initial count - should be equal to 1000
+        //Use collection getSize in order to get quick result
+        $this->assertEquals(1000, $this->quoteCollectionFactory->create()->getSize());
+
+        //Deleting expired quotes
+        $this->cleanExpiredQuotes->execute();
+        $totalCount = $this->quoteCollectionFactory->create()->getSize();
+
+        //There should be no quotes anymore
+        $this->assertEquals(0, $totalCount);
```
Comment (lbajsarowicz) [inline dev/tests/integration/testsuite/Magento/Sales/Cron/CleanExpiredQuotesTest.php]: Extract that to private method `assertQuotesCount(int $expectedCount)`

File: dev/tests/integration/testsuite/Magento/Sales/Cron/CleanExpiredQuotesTest.php [hunk 2]
```diff
@@ -66,4 +73,24 @@ public function testExecute()
...
+     * Check if outdated quotes are deleted.
+     *
+     * @magentoConfigFixture default_store checkout/cart/delete_quote_after -365
+     * @magentoDataFixture Magento/Sales/_files/quotes_big_amount.php
+     */
+    public function testExecuteWithBigAmountOfQuotes()
+    {
+        //Initial count - should be equal to 1000
+        //Use collection getSize in order to get quick result
+        $this->assertEquals(1000, $this->quoteCollectionFactory->create()->getSize());
```
Comment (lbajsarowicz) [inline dev/tests/integration/testsuite/Magento/Sales/Cron/CleanExpiredQuotesTest.php]: `$this->assertQuotesCount(1000)`

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Improve readability of Integration Tests, and we're :heavy_check_mark:

**Review by lbajsarowicz — APPROVED** [review]
:heavy_check_mark: Great! Thank you for your contribution!

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7193](https://jira.corp.magento.com/browse/ENGCOM-7193) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

Tested with ~ 10000 expired quotes

Run: `php bin/magento cron:run`

**Before:**
Screenshot from 2020-04-03 16-44-56

**After:**
Screenshot from 2020-04-03 16-59-52

**Comment by ihor-sviziev** [comment]
So maybe it’s better to adjust my integration test with bigger number of
quotes? Could you let me know from which amount it started failing, I’ll
just add it?

**Comment by engcom-Alfa** [comment]
@ihor-sviziev in my case, it started failing with 5000 expired quotes and more, with memory_limit=128M

**Comment by ihor-sviziev** [comment]
@engcom-Alfa for now let's leave it as is, as increasing amount of quotes might cause fatal error in integration tests due to out of memory. Why - because there running also other tests that could leave some not cleaned memory

**Comment by sdzhepa** [comment]
@magento create issue

**Comment by engcom-Echo** [comment]
Failed functional tests not related to the changes in this PR

**Pull request merged**